src/Classes/Utils/CanvasUtils.ts

```diff
@@ -0,0 +1,348 @@
+        //     delete ctx.shadowOffsetY;
+        // }
+
+        context.fillStyle = colorValue;
+
+        const pos = {
+            x: particle.position.x,
+            y: particle.position.y,
+        };
+
+        context.translate(pos.x, pos.y);
+        context.beginPath();
+
+        if (particle.angle !== 0) {
+            context.rotate(particle.angle * Math.PI / 180);
+        }
+
+        if (backgroundMask) {
+            context.globalCompositeOperation = 'destination-out';
+        }
+
+        CanvasUtils.drawShape(context, particle, radius, stroke);
+
+        context.closePath();
+
+        if (stroke.width > 0) {
+            context.strokeStyle = stroke.color;
+            context.lineWidth = stroke.width;
+            context.stroke();
+        }
+
+        context.fill();
+        context.restore();
+    }
+
+    private static drawShape(context: CanvasRenderingContext2D, particle: Particle, radius: number, stroke: IStroke): void {
+        const pos = {
+            x: particle.offset.x,
+            y: particle.offset.y,
+        };
+
+        switch (particle.shape) {
+            case ShapeType.line:
+                CanvasUtils.drawLineShape(context, radius, stroke);
+                break;
+
+            case ShapeType.circle:
+                CanvasUtils.drawCircleShape(context, radius, pos);
+                break;
+
+            case ShapeType.edge:
+            case ShapeType.square:
+                CanvasUtils.drawSquareShape(context, radius);
+                break;
+
+            case ShapeType.triangle:
+                CanvasUtils.drawTriangleShape(context, radius);
+                break;
+
+            case ShapeType.polygon:
+                CanvasUtils.drawPolygonShape(context, radius, particle.container.options.particles.shape.polygon.sides);
+                break;
+
+            case ShapeType.star:
+                CanvasUtils.drawStarShape(context, radius, particle.container.options.particles.shape.polygon.sides);
+                break;
+
+            case ShapeType.heart:
+                CanvasUtils.drawHeartShape(context, radius);
+                break;
+
+            case ShapeType.char:
+            case ShapeType.character:
+                const style = particle.container.options.particles.shape.character.style;
+                const weight = particle.container.options.particles.shape.character.weight;
+                const size = Math.round(radius) * 2;
+                const font = particle.container.options.particles.shape.character.font;
+                const text = particle.text;
+
+                context.font = `${style} ${weight} ${size}px ${font}`;
+
+                if (text) {
+                    const pos = {
+                        x: -radius / 2,
+                        y: radius / 2
+                    };
+
+                    CanvasUtils.drawTextShape(context, text, pos, particle.container.options.particles.shape.character.fill);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Exceeds maximum line length of 120](https://app.codacy.com/manual/ar3s/tsparticles/pullRequest?prid=5171185)